### DISCUSSION TITLE
Support repro PCB for reflashing 3v and 5v SMT flash chip

### DISCUSSION THREAD

**Discussion by StanMOI** [sanni/cartreader#116] [Need help]
Hi Sanni,

You said here :
https://github.com/sanni/cartreader/wiki/Reflashing-SNES-Repros
that some of Chinese PCB repro can be reflashed
I see on your picture flash chip 29LV320 witch use 3v
I know that all Modern flash is 3V and will require voltage translation.
Supply 3.3V flash with 3.3V and interface it to a higher voltage without translation, it'll damage chip.
You put list of all supported flash, they are all 3v supply:
AM29F016/AM29F032
MX29F1610
MX29L3211
MX29LV160
MX29LV320
S29GL032
Do they all supported only to be flashed with adapters ? 
Or on SNES repro PCB also?

Here is my next questions:
- I want to be sure that your Cartreader support to reflash SNES PCB Repro with 3v supply flash chips ?
- Do your Cartreader support reflashing SNES PCB with chips like 29F033 (5v supply)?
- Is it possible to reflash SNES PCB Repro with chips like M29W128GH ?
- Is it be possible to reflash SNES PCB Repro with bigger chips capacity like :
S29GL512 (with is 512Mbit = 64Mo) or S29GL01G (with is 1Gbit = 128Mo), both 3v supply ?

Thanks

**Comment by sanni** [comment]
The flash adapter follows the SNES HiROM pinout so I'm pretty sure that all the listed flashroms can also be flashed from the SNES Repro menu. I only tried with 29F033 and 29LV320 though. But they all *should* work.

Now ofc it will depend on how the repro PCB is designed since you will need access to the flashroms WE pin. Thats what the wire on the 29LV320 repro is for. 

The SNES is a 5V system so all repro PCBs should be designed to work with 5V. Thats why I flashed the 29LV320 with 5V in the Wiki article.
The Cart Reader itself can be switched between 3V and 5V but the repro PCB has diodes that drop the voltage from 5V to 3V so flashing with 3V didn't work in this case because the diodes on the repro PCB would drop the voltage to 1.6V or so.

Flashing 29F033 repros works fine.

Possible if you add support for the M29W128GH chip yourself, you will need to read the datasheet and then add a new function. I have decided to not work on flashing flashroms/repros anymore. So it would be up to you.

Same answer as before, possible yes, but you will have to write the needed function yourself with no help from me.

**Reply by StanMOI** [reply]
> The flash adapter follows the SNES HiROM pinout so I'm pretty sure that all the listed flashroms can also be flashed from the SNES Repro menu. I only tried with 29F033 and 29LV320 though. But they all _should_ work.
> 
> Now ofc it will depend on how the repro PCB is designed since you will need access to the flashroms WE pin. Thats what the wire on the 29LV320 repro is for.
> 
> The SNES is a 5V system so all repro PCBs should be designed to work with 5V. Thats why I flashed the 29LV320 with 5V in the Wiki article.
> The Cart Reader itself can be switched between 3V and 5V but the repro PCB has diodes that drop the voltage from 5V to 3V so flashing with 3V didn't work in this case because the diodes on the repro PCB would drop the voltage to 1.6V or so.
> 
> Flashing 29F033 repros works fine.
> 
> Possible if you add support for the M29W128GH chip yourself, you will need to read the datasheet and then add a new function. I have decided to not work on flashing flashroms/repros anymore. So it would be up to you.
> 
> Same answer as before, possible yes, but you will have to write the needed function yourself with no help from me.

« since you will need access to the flashroms WE pin. »
Where can I find information to how PCB most be design to make it compatible with your reader?

**Comment by StanMOI** [comment]
You say « that’s why I flashed 29LV320 with 5v »
On picture I see that you switch on 3v 
![image](https://user-images.githubusercontent.com/38383423/113481404-b1461780-9499-11eb-810a-44929d7a5e2e.jpeg)

**Comment by sanni** [comment]
If it's just the chip you use 3V. If the chip is soldered to a SNES repro you use 5V.

**Reply by StanMOI** [reply]
I see better now.
- if flash chip 3v is soldered on PCB = switch to 5v ( cause PCB have diodes and they will drop volage to 3v)

But,
when it’s 5v flash chip on PCB?
And you will switch to 5v, one its pass through diodes, Will voltage be enough to flash chip ?

**Comment by sanni** [comment]
The repros with 5V flashrom chips don't have a diode on the VCC pin of the flashrom, so the full 5V will pass through.

**Comment by StanMOI** [comment]
To make support repro PCB with other flash chip (like S29GL512), file I have to edit is "FLASH.ino" ?  
And the code line should be like this one, yes?

```
else if ((strcmp(flashid, "C2C9") == 0) || (strcmp(flashid, "C2CB") == 0)) {
    println_Msg(F("MX29LV640 detected"));
    println_Msg(F("ATTENTION 3.3V"));
    flashSize = 8388608;
    flashromType = 2;
  }
```
<img width="773" alt="Capture d’écran 2021-04-13 à 03 13 54" src="https://user-images.githubusercontent.com/38383423/114482297-a8c2af00-9c06-11eb-8fe9-a26be06b32e5.png">

"C2C9" :  means Top Boot sector
C2 = Manufacturer ID (DATA HEX)
C9 = Divice ID (DATA HEX)

"C2CB" :  means Bottom Boot sector
C2 = Manufacturer ID (DATA HEX)
CB = Divice ID (DATA HEX)

Is that right ?

If yes, for S29GL512 its should looks like that:
```
else if ((strcmp(flashid, "DQ7") == 0) || (strcmp(flashid, "DQ0") == 0)) {
    println_Msg(F("S29GL512 detected"));
    println_Msg(F("ATTENTION 3.3V"));
    flashSize = 33554432;
    flashromType = 2;
  }
```
<img width="763" alt="Capture d’écran 2021-04-13 à 03 44 50" src="https://user-images.githubusercontent.com/38383423/114484507-a3fffa00-9c0a-11eb-9050-ca9571b5a0a8.png">

That all I have to add?

**Comment by sanni** [comment]
The Flashid should be 017E and 23, and needs to be added to the code starting line 498 in flash.ino like so:

```
 else if (strcmp(flashid, "017E") == 0) {    
    // S29GL032M   
    if (readByte_Flash(28) == 0x1A) {    
      println_Msg(F("S29GL032M detected"));    
      flashSize = 4194304;    
      sectorSize = 65536;    
      bufferSize = 32;    
    }   
 else if (readByte_Flash(28) == 0x23) {    
      println_Msg(F("S29GL512N detected"));    
      flashSize = 67108864;    
      sectorSize = 131072;    
      bufferSize = 32;    
    }   
    // Unknown S29GL type    
    else {    
      println_Msg(F("Unknown S29GL Type"));    
      flashSize = 4194304;    
      sectorSize = 65536;    
      bufferSize = 32;    
    }    
    println_Msg(F("ATTENTION 3.3V"));    
    flashromType = 2;    
  }    
```

Then you have to find out if there are any other differences between the already supported S29GL type and your flashrom
S29GL512N: https://www.cypress.com/file/219941/download
S29GL032M: https://datasheetspdf.com/pdf-file/1067559/SPANSION/S29GL032M/1

and make further changes to writeFlash29GL line 1304:

```
/******************************************
  S29GL flashrom functions
*****************************************/
void writeFlash29GL(unsigned long sectorSize, byte bufferSize) {    
  // Create filepath    
  sprintf(filePath, "%s/%s", filePath, fileName);    
  println_Msg(F("Flashing file "));    
  println_Msg(filePath);    
  display_Update();    

  // Open file on sd card    
  if (myFile.open(filePath, O_READ)) {    
    // Get rom size from file   
    fileSize = myFile.fileSize();    
    if (fileSize > flashSize)    
      print_Error(F("File size exceeds flash size."), true);    

    // Set data pins to output    
    dataOut();   

    for (unsigned long currSector = 0; currSector < fileSize; currSector += sectorSize) {    
      // Blink led   
      PORTB ^= (1 << 4);    

      // Write to flashrom    
      for (unsigned long currSdBuffer = 0; currSdBuffer < sectorSize; currSdBuffer += 512) {   
        // Fill SD buffer    
        myFile.read(sdBuffer, 512);   

        // Write bufferSize bytes at a time    
        for (int currWriteBuffer = 0; currWriteBuffer < 512; currWriteBuffer += bufferSize) {   
          // 2 unlock commands    
          writeByte_Flash(0x555 << 1, 0xaa);   
          writeByte_Flash(0x2aa << 1, 0x55);    
          // Write buffer load command at sector address   
          writeByte_Flash(currSector + currSdBuffer + currWriteBuffer, 0x25);    
          // Write byte count (minus 1) at sector address   
          writeByte_Flash(currSector + currSdBuffer + currWriteBuffer, bufferSize - 1);   

          // Load bytes into buffer
          for (byte currByte = 0; currByte < bufferSize; currByte++) {
            writeByte_Flash(currSector + currSdBuffer + currWriteBuffer + currByte, sdBuffer[currWriteBuffer + currByte]);
          }

          // Write Buffer to Flash      
          writeByte_Flash(currSector + currSdBuffer + currWriteBuffer + bufferSize - 1, 0x29);   

          // Read the status register at last written address    
          dataIn8();     
          byte statusReg = readByte_Flash(currSector + currSdBuffer + currWriteBuffer + bufferSize - 1);     
          while ((statusReg & 0x80) != (sdBuffer[currWriteBuffer + bufferSize - 1] & 0x80)) {     
            statusReg = readByte_Flash(currSector + currSdBuffer + currWriteBuffer + bufferSize - 1);    
          }    
          dataOut();      
        }   
      }    
    }    
    // Set data pins to input again    
    dataIn8();   
    // Close the file:   
    myFile.close();   
  }    
  else {      
    println_Msg(F("Can't open file on SD"));    
    display_Update();   
  }   
}
```

Basically a lot of reading the datasheet and then trial&error until it works.

**Comment by Gilberto99** [comment]
@sanni Hello friend, do you have the Gerbers for this 29lv320 adapter for me to make snes repros? I would be very grateful
